src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewFactory.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Responsible for wiring up all the different components of the document outline feature.
+    /// </summary>
+    internal static class DocumentOutlineViewFactory
+    {
+        public static DocumentOutlineView CreateView(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            threadingContext.ThrowIfNotOnUIThread();
+            var visualStudioCodeWindowInfoService = new VisualStudioCodeWindowInfoService(codeWindow, editorAdaptersFactoryService, threadingContext);
+            var textViewEventSource = CreateEventSource(asyncListener, visualStudioCodeWindowInfoService);
+            var viewModel = new DocumentOutlineViewModel(languageServiceBroker, asyncListener, visualStudioCodeWindowInfoService, textViewEventSource);
```

Comment:
some of these terms are a bit weird for me.  In general, i expect a model to be pure data and otherwise no functionality.  But this model is taking in an event source.  In an MVC system the model is the pure data, the view presents the data, and the Controller is the one that manages things like events, updating the model, and then bringing the view in line with hte model.

Reply:
[Model-View-ViewModel](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel) is the standard WPF way of doing user interfaces. These typically map to the MVC pattern as follows:

- Model -> Model
- View -> View
- Controller ~> ViewModel

A ViewModel isn't really a controller but instead is a component that has no thread affinity and has its data _bound_ to parts of the UI. This allows us to update data from a background thread and let the WPF data-binding system handle updating the UI in an efficient manner.